tinygrad/tensor.py

```diff
@@ -418,21 +418,21 @@ def calc_dim(tensor_dim:int) -> int:
 
       # track tensor_dim and tensor_index using a dict
       # calc_dim to get dim and use that to normalize the negative tensor indices
-      idx: Dict[int,Tensor] = {(dim := calc_dim(td)):(tensor<0).where(ret.shape[dim],0) + tensor for td,tensor in zip(type_dim[Tensor],tensor_index)}
-
-      # compute sum_dim, arange, and idx
-      max_idx_dim, first_dim, last_dim = max(i.ndim for i in idx.values()), min(idx.keys()), max(idx.keys())
-      sum_dim = tuple(d if n==0 else d+max_idx_dim-n for n,d in enumerate(idx.keys()))
-      arange = [Tensor.arange(ret.shape[d], requires_grad=False, device=self.device).reshape(ret.shape[d], *[1]*(ret.ndim+max_idx_dim-n-sd-1)) \
-                for n,(sd,d) in enumerate(zip(sum_dim, idx.keys()))]
-      reshaped_idx = [i.reshape(i.shape + (1,)*(ret.ndim - first_dim - (n or 1))) for n,i in enumerate(idx.values())]
-      ret = ret.reshape(ret.shape[:first_dim+1] + (1,)*max_idx_dim + ret.shape[first_dim+1:])
-
-      # iteratively eq -> mul -> sum fancy index
-      try:
-        for a,i,sd in zip(arange, reshaped_idx, sum_dim): ret = (a==i).mul(ret).sum(sd)
+      idx: Dict[int,Tensor] = {(dim := calc_dim(td)):(tensor<0).where(ret.shape[dim],0) + tensor for td,tensor in zip(type_dim[Tensor], tensor_index)}
+
+      eqs, first_dim, max_idx_dim, last_dim  = [], min(idx.keys()), max(i.ndim for i in idx.values()), max(idx.keys())
+      for dim, i in idx.items():
+        a = i.reshape(i.shape + (1,)*(ret.ndim - first_dim))
+        b = Tensor.arange(ret.shape[dim]).reshape(ret.shape[dim:dim+1] + (1,)*(ret.ndim - dim - 1))
+        eqs.append(a == b)
+
+      try: mask = reduce(lambda x,y: x.mul(y), eqs).contiguous()
```

Comment:
this `contiguous` is cuz maybe a bug in the shapetrackers
or maybe view not merging 
or idk

gotta to some investigating 🧐

Reply:
Yea, that needs to be understood. contiguous should never change the final output (we should fuzz this)